doc/development/adding_operators.rst

```diff
@@ -0,0 +1,347 @@
+specific subclasses.
+
+* Operators inheriting from :class:`~.Observable` support addition and scalar multiplication:
+
+  >>> op = qml.PauliX(0) + 0.1 * qml.PauliZ(0)
+  >>> op.name
+  Hamiltonian
+  >>> op
+    (0.1) [Z0]
+  + (1.0) [X0]
+
+* Operators may define a hermitian conjugate:
+
+  >>> qml.RX(1., wires=0).adjoint()
+  RX(-1.0, wires=[0])
+
+Creating custom operators
+#########################
+
+A custom operator can be created by inheriting from :class:`~.Operator` or one of its subclasses.
+
+The following is an example for a custom gate that possibly flips a qubit and then rotates another qubit.
+The custom operator defines a decomposition, which the devices can use (since it is unlikely that a device
+knows a native implementation for ``FlipAndRotate``). It also defines an adjoint operator.
+
+.. code-block:: python
+
+    import pennylane as qml
+
+
+    class FlipAndRotate(qml.operation.Operation):
+
+        # Define how many wires the operator acts on in total.
+        # In our case this may be one or two, which is why we
+        # use the AnyWires Enumeration to indicate a variable number.
+        num_wires = qml.operation.AnyWires
+
+        # This attribute tells PennyLane what differentiation method to use. Here
+        # we request parameter-shift (or "analytic") differentiation.
+        grad_method = "A"
+
+        def __init__(self, angle, wire_rot, wire_flip=None, do_flip=False,
+                           do_queue=True, id=None):
+
+            # checking the inputs --------------
+
+            if do_flip and wire_flip is None:
+                raise ValueError("Expected a wire to flip; got None.")
+
+            # note: we use the framework-agnostic math library since
+            # trainable inputs could be tensors of different types
+            shape = qml.math.shape(angle)
+            if len(shape) > 1:
+                raise ValueError(f"Expected a scalar angle; got angle of shape {shape}.")
+
+            #------------------------------------
+
+            # do_flip is not trainable but influences the action of the operator,
+            # which is why we define it to be a hyperparameter
+            self._hyperparameters = {
+                "do_flip": do_flip
+            }
+
+            # we extract all wires that the operator acts on,
+            # relying on the Wire class arithmetic
+            all_wires = qml.wires.Wires(wire_rot) + qml.wires.Wires(wire_flip)
+
+            # The parent class expects all trainable parameters to be fed as positional
+            # arguments, and all wires acted on fed as a keyword argument.
+            # The id keyword argument allows users to give their instance a custom name.
+            # The do_queue keyword argument specifies whether or not
+            # the operator is queued when created in a tape context.
+            super().__init__(angle, wires=all_wires, do_queue=do_queue, id=id)
+
+        @property
+        def num_params(self):
```

Comment:
Please tell developers here that a static property is preferred if num_params is the same for each instance. This also means less typing...